File: Assets/user/jwu/script/weapon/infrared.js
// ======================================================================================
// File         : Infrared.js
// Author       : Wu Jie 
// Last Change  : 08/25/2010 | 22:42:29 PM | Wednesday,August
// Description  : 
// ======================================================================================

///////////////////////////////////////////////////////////////////////////////
// properties
///////////////////////////////////////////////////////////////////////////////

var material: Material;
var lineSize = 0.2;
var lineColor : Color;

///////////////////////////////////////////////////////////////////////////////
// function define
///////////////////////////////////////////////////////////////////////////////

// ------------------------------------------------------------------ 
// Desc: 
// ------------------------------------------------------------------ 

function Start () {
    var lr = gameObject.GetComponent(LineRenderer);
    if ( lr == null ) {
        lr = gameObject.AddComponent(LineRenderer);
        lr.useWorldSpace = false;
        lr.SetWidth(lineSize, lineSize);
        lr.material = material;
        lr.SetPosition(0, Vector3.zero );
        lr.SetPosition(1, Vector3.forward * 10.0 );
        lr.SetColors(lineColor,lineColor);
    }
}

// ------------------------------------------------------------------ 
// Desc: 
// ------------------------------------------------------------------ 

function Update () {
    var lr = GetComponent(LineRenderer);
    if ( lr ) {

        var layerMask = 1 << 2;
        // This would cast rays only against colliders in layer 2.
        // But instead we want to collide against everything except layer 2. 
        // The ~ operator does this, it inverts a bitmask.
        layerMask = ~layerMask;

        var dist = 100.0;
        var hit : RaycastHit;
        var fwd = transform.TransformDirection (Vector3.forward);
        if ( Physics.Raycast (transform.position, fwd, hit, 100.0, layerMask ) ) {
            dist = hit.distance;
        }
        lr.SetPosition( 1, Vector3.forward * dist );
    }
}

File: Assets/user/jwu/script/weapon/Infrared.cs
// ======================================================================================
// File         : Infrared.cs
// Author       : Wu Jie 
// Last Change  : 09/27/2010 | 21:53:05 PM | Monday,September
// Description  : 
// ======================================================================================

///////////////////////////////////////////////////////////////////////////////
// usings
///////////////////////////////////////////////////////////////////////////////

using UnityEngine;
using System.Collections;

///////////////////////////////////////////////////////////////////////////////
// defines
///////////////////////////////////////////////////////////////////////////////

public class Infrared : MonoBehaviour {

    ///////////////////////////////////////////////////////////////////////////////
    // properties
    ///////////////////////////////////////////////////////////////////////////////

    public Material material;
    public float lineSize = 0.2f;
    public Color lineColor;

    ///////////////////////////////////////////////////////////////////////////////
    // functions
    ///////////////////////////////////////////////////////////////////////////////

    // ------------------------------------------------------------------ 
    // Desc: 
    // ------------------------------------------------------------------ 

    void Start () {
        LineRenderer lr = GetComponent(typeof(LineRenderer)) as LineRenderer;
        if ( lr == null ) {
            lr = gameObject.AddComponent(typeof(LineRenderer)) as LineRenderer;
            lr.useWorldSpace = false;
            lr.SetWidth(lineSize, lineSize);
            lr.material = material;
            lr.SetPosition(0, Vector3.zero );
            lr.SetPosition(1, Vector3.forward * 10.0f );
            lr.SetColors(lineColor,lineColor);
        }
    }

    // ------------------------------------------------------------------ 
    // Desc: 
    // ------------------------------------------------------------------ 

    void Update () {
        LineRenderer lr = GetComponent(typeof(LineRenderer)) as LineRenderer;
        if ( lr ) {
            // This would cast rays only against colliders in layer x.
            // ignore layer: bullet_player, player, trigger
            int layerMask = 1 << Layer.bullet_player 
                | 1 << Layer.player 
                | 1 << Layer.trigger;

            // But instead we want to collide against everything except layer x. 
            // The ~ operator does this, it inverts a bitmask.
            layerMask = ~layerMask;

            float dist = 100.0f;
            RaycastHit hit;
            Vector3 fwd = transform.TransformDirection (Vector3.forward);
            if ( Physics.Raycast (transform.position, fwd, out hit, 100.0f, layerMask ) ) {
                dist = hit.distance;
            }
            lr.SetPosition( 1, Vector3.forward * dist );
        }
    }
}
